### ISSUE TITLE
Imported variable bindings have difficult to debug names.

### ISSUE THREAD

**Issue by everett1992** [webpack/webpack#3957]
**Do you want to request a *feature* or report a *bug*?**
Feature, improvement to debugging webpack-built projects.

**What is the current behavior?**
Webpack renames variable bindings in  ES6 imports when it replaces them with `__webpack_require__` calls.  In this example the imported binding `foo` is compiled to `__WEBPACK_IMPORTED_MODULE_0__foo__["a" /* foo */]`. If you use dev tools debugger you cannot watch or execute`foo` because it is not defined. 

input:
```javascript
import {foo} from './foo';
console.log(foo());
```
output: 
```javascript
Object.defineProperty(exports, "__esModule", { value: true });                 
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0__foo__ = __webpack_require__(0);                                                                          
console.log(__webpack_require__.i(__WEBPACK_IMPORTED_MODULE_0__foo__["a" /* foo */])());
```
Debugging is difficult, you can't watch or reference the original binding. This is made *worse* by sourcemaps because you can't see new name in the original source. It's not immediately apparent why a line copied from the current context into the console results in `not defined` errors. Chrome even removes code that has a corresponding source map, so it becomes difficult to find the in-code name of an imported value.

**If the current behavior is a bug, please provide the steps to reproduce.**
[Simple reproduction](https://gist.github.com/everett1992/f5aa2a82641e220baf74923c962ef102).  

**What is the expected behavior?**
Variable bindings should be preserved by name.

**If this is a feature request, what is motivation or use case for changing the behavior?**
Debugging a project built by webpack today is more difficult than it aught to be.

**Comment by sokra** [comment]
> Variable bindings should be preserved by name.

If you tell me how this is possible, I'll do it.

**Comment by everett1992** [comment]
I'm not familiar with webpack's internals, or intricacies of the ES6 spec, so I understand why this could be harder than an outside observer would think.

Why can't the example in the issue be compiled to

```javascript
Object.defineProperty(exports, "__esModule", { value: true });                 
/* harmony import */ var foo = __webpack_require__(0)["a" /* foo */];                                                                          
console.log(__webpack_require__.i(foo)());
```

I guess I see the problem, if the module  `./foo` were to redefine it's export `foo` the binding in `index.js` would be out of date.

1. Could one use static-analysis to check for happy-case no export redefinition?
2. Could ES6 proxies help here (Proxy support probably prevents this from being a viable choice)

**Comment by sokra** [comment]
> I guess I see the problem, if the module ./foo were to redefine it's export foo the binding in index.js would be out of date.

yes, that's the problem. ES import are live bindings. Actually the `a` property in `__webpack_require__(0)` is a getter that may return different values each time.

**Comment by Hanterdro** [comment]
Yeah, this is a problem for me, too. Debugging is much harder now :-(

**Comment by olsonpm** [comment]
Regarding whether it's possible, babel transforms exported binding assignments to export assignments directly. e.g.

```js
// from
export let a = 5;
export function setAto10() {
  a = 10 <------- from a simple variable assignment
}

// to
exports.setAto10 = setAto10;
var a = exports.a = 5;
exports.a = a;
function setAto10() {
  exports.a = a = 10 // <------- to an exported variable assignment
}
```

This solves the issue and avoids the need to use a getter.  I can't think of any reason why the approach wouldn't work with webpack however I've only been researching this problem as of tonight.

**Comment by sokra** [comment]
@olsonpm This issue is about imported variables not exported.

The only way which I see to fix it is:

``` js
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0__foo__ = __webpack_require__(0);
with({
  get foo() { return __WEBPACK_IMPORTED_MODULE_0__foo__ ["a"]; }
}) {
  // you code
  // debugger could access `foo`
}
```

But... no...

**Comment by sdwio** [comment]
One small step could be to preserve the variable names in the  `__WEBPACK_IMPORTED_MODULE_0__foo__` array. In the example it is not that obvious since the file and variable are named foo. If you have `import {bar, qux, sol} from './foo';` the three variables will end up as "a", "b", and "c". If they would be visible as `__WEBPACK_IMPORTED_MODULE_0__foo__["bar"]` etc. it would still not be perfect but usable.
Yet I have no idea if this might cause other complications.

**Comment by yang** [comment]
For projects that don't care for the live binding requirement and do care about a **sane interactive debugging experience**, can there be an option to just treat imports like regular bindings?  (I imagine there are many such projects.)

**Comment by valscion** [comment]
Hmmm... what if webpack just flipped the variable name to be _first_ and then appends the `__WEBPACK_IMPORTED_MODULE_X` stuff to the end? That way at least you could try to find the variable with tab autocompletion?

input:

```js
import {foo} from './foo';
console.log(foo());
```

output:

```js
Object.defineProperty(exports, "__esModule", { value: true });                 
/* harmony import */ var __foo__WEBPACK_IMPORTED_MODULE_0 = __webpack_require__(0);                                                                          
console.log(__webpack_require__.i(__foo__WEBPACK_IMPORTED_MODULE_0["a" /* foo */])());
```

Then you could write `__foo` and locate the correct variable from the autocomplete result.

**Comment by thecotne** [comment]
@sokra that is actually very good for development (i think you also need set function to preserve all behaviors ...)

and for production you will stick with current version (so uglify can do his things)

am i missing something?

**Comment by thecotne** [comment]
i did some research and i am missing strict mode 

but you can go around this issue by something like this

```js
(() => {
  const scope = {
    get name() { return 'cotne' }
  }
  // strict mode does not apply here so we can use "with" statement 
  with (scope) {
    (() => {
      "use strict"
      // package code goes here where strict mode applies 
      // you can't use "with" statement here but outer one is fine
      console.log(name)
    })()
  }
})()

```

**Comment by MagicDuck** [comment]
@sokra: stupid question. Could we just add an option to webpack which will make it assume that the harmony bindings are not live when turned on. Maybe somebody could correct me, but I can't think of any code where live bindings would be really useful/be a good practice. At least in our code base, I believe this to be the case. To me, it contradicts the whole concept ES6 import/export declaration being static. Or would this introduce a problem with stuff from node_modules/? Although in that case we can have some sort of regex to determine which modules to do this for...

**Comment by gaearon** [comment]
I don’t think suggesting to treat modules not per spec is a good idea. I wouldn’t like to see an option to disable live bindings, it’s only going to cause endless confusion.

Rollup somehow is able to preserve names in live bindings (due to module concatenation). I doubt that would be fast enough for development, but thought it’s worth mentioning.

**Comment by MagicDuck** [comment]
@gaearon and the weird import names are not causing endless confusion...? 😆

**Comment by gaearon** [comment]
They are a debugging inconvenience, sure. (Potentially solvable by tooling.)

Coming up with your own language semantics to avoid an inconvenience is a much harder problem to unravel.

**Comment by MagicDuck** [comment]
@gaearon I understand your point of view and I think it would be ideal, but on the other hand I am supporting hundreds of devs which often complain about this problem. What am I to tell them, suck it up and get used to adding the magic incantation before your vars ? The other thing is, I have been reading that one of the reasons for the live nature of module bindings is circular devs. Would that not mean the current as black implementation is already broken, or am I missing something?

**Comment by MagicDuck** [comment]
Ugh autocorrect. Deps not devs at the end there. And webpack not "as black " lol

**Comment by sokra** [comment]
`__webpack_require__.i` no longer exists.

@thecotne The code you posted actually looks promising. I guess we can use it. It looks ugly but you won't see it with Source Maps.

**Comment by MagicDuck** [comment]
I will add a PR with @thecotne 's suggestion if there is no objection...

**Comment by NikhilVerma** [comment]
Does the source map API support a way to bind variable names to functions?

e.g. If I try to watch "foo" it will internally run ```return __WEBPACK_IMPORTED_MODULE_0__foo__ ["a"];```

**Comment by NikhilVerma** [comment]
One approach I am considering to try is to use babel loaders during development, they will convert imports to requires so it will keep development easy

**Comment by olee** [comment]
I also think @valscion's idea to put the variable name to the front would at least help a bit to make it easier to find the correct import variable.
Also what do you mean @sokra with `__webpack_require__.i` no longer exists?
I see it in my current builds as well with iirc up to date webpack (I might be wrong though).

**Comment by MagicDuck** [comment]
Sorry for the non existent PR so far..
I started working on it, realized it was a bit trickier than anticipated, and then life came in the way ☹

**Comment by thecotne** [comment]
if we can't use `with` statement something like this also will be much better then what we have now

```js
(() => {
  "use strict"

  const imported = {
    get name() { return 'cotne' }
  }

  console.log(imported.name)
})()
```

**Comment by epikhighs** [comment]
Since the problem lies with ES6 modules, couldn't there be a way to "simply" covert the ES6 module to CommonJS, then have webpack load the CommonJS module?  It would be a viable workaround for development as long as I could debug these ES6 modules.

**Comment by hedgepigdaniel** [comment]
Adding `transform-es2015-modules-commonjs-simple` might solve this problem, but it creates a much bigger problem: There will be no error if you import an identifier from an ES6 module which is not exported by it. There will also be no tree shaking.

That is an absolute disaster. I can't count the hours I've wasted debugging a problem because the compiler didn't tell me that I imported something that didn't exist. Or the live bugs I've released because when I changed an exported name I forgot to change one of the imports.

**Comment by richardaum** [comment]
Nothing here besides `transform-es2015-modules-commonjs-simple` ?

**Comment by staeke** [comment]
Here's another suggestion then:

Rather than going with a `with` block, we could wrap the whole file with a IIFE with named parameters. That would work with debugging the 99% case when exports are constant, while still allowing the actual code run as defined in ES6. Also, this would only apply if in dev mode.

This:
```js
import {a, b} from 'foo'
console.log(a);
console.log(b);
```

would then be transformed to this (I've omitted the hideous WEBPACK_IMPORTED_MODULE...)
```js
var __foo = __webpack_require__(0)
((a, b) => {
    console.log(__foo["a"]);
    console.log(__foo["b"]);
})(__foo["a"], __foo["b"]);
```

**Comment by NikhilVerma** [comment]
> Here's another suggestion then:
> 
> Rather than going with a `with` block, we could wrap the whole file with a IIFE with named parameters. That would work with debugging the 99% case when exports are constant, while still allowing the actual code run as defined in ES6. Also, this would only apply if in dev mode.
> 
> This:
> 
> ```js
> import {a, b} from 'foo'
> console.log(a);
> console.log(b);
> ```
> 
> would then be transformed to this (I've omitted the hideous WEBPACK_IMPORTED_MODULE...)
> 
> ```js
> var __foo = __webpack_require__(0)
> ((a, b) => {
>     console.log(__foo["a"]);
>     console.log(__foo["b"]);
> })(__foo["a"], __foo["b"]);
> ```

If this has performance implications it can only be enabled for ```development``` mode

**Comment by Gvozd** [comment]
@sokra , May support mutated imports, if register setter in consumer module, as the getter is now registered in the source module?
Sоmething like this:

consumer.js
```js
import {foo as bar} from './source';
setInterval(() => console.log(bar), 1000);
// -->

// when changed(=, +=, ...) export `foo` in module `sourceModuleId`, then need call setters of all consumer modules
__webpack_require__.registerSetter(__webpack_exports__, sourceModuleId, "foo", function(val) { bar = val; })
let bar;
setInterval(() => console.log(bar), 1000);//no changes. imported variable available as is
```

source.js:
```js
export let foo = 123;
setInterval(() => {
  foo = foo + 1;
}, 1000);
// ---->
let foo;
__webpack_require__.notify(sourceModuleId, 'foo', foo = 123);
setInterval(() => {
  //exported variable available in scope as is
  __webpack_require__.notify(sourceModuleId, 'foo', foo = foo + 1);
}, 1000);
```

Of course, this solution can lead to performance problems, but this behavior has the right to live in development mode.

**Comment by Moghul** [comment]
If you guys don't mind me asking, what is the debugging process supposed to look like with imports like this? How do you track down "unexpected behaviour" style bugs if you can't easily identify what you're looking at and what the values are? This is almost like debugging obfuscated code.

**Comment by NikhilVerma** [comment]
> If you guys don't mind me asking, what is the debugging process supposed to look like with imports like this? How do you track down "unexpected behaviour" style bugs if you can't easily identify what you're looking at and what the values are? This is almost like debugging uglified/obfuscated code.

It's not convenient at all. my strategies are:

1. **Disable source maps.** Then you can see the real variable names and debug them properly.
2. **Memorise the webpack naming conventions.** If you are at a breakpoint, and your import is similar to ```import UserModel from 'xxx'```. Then the webpack bindings are usually called ```__WEBPACK_IMPORTED_MODULE_0__UserModel__```. The devtools are usually friendly enough to autocomplete for you if you start typing ```__``` and then go from there.

**Comment by Moghul** [comment]
Ok so what you're saying is "use commonjs".

**Comment by Yehonal** [comment]
I think it's an issue with all kind of transpilers: https://github.com/microsoft/TypeScript/issues/9627#issuecomment-643192125

**Comment by staeke** [comment]
For webpack 4, use this plugin. This will simplify names if you're using `eval` and allow straight import vars to be used with source maps. 

So this code:

```js
import a, {b, c as d} from 'foo'
a()
```

becomes 

```js
var __foo = webpack_require('...'); let {default: a, b, c: d} = __foo; if (typeof d === "undefined") d = _foo
_foo.a()
```

https://gist.github.com/staeke/f197ff144e379aa884f91b95003f4ee1

**Comment by epikhighs** [comment]
I couldn't get `transform-es2015-modules-commonjs-simple` to work anymore, but setting `modules: "cjs"` in babel config's `@babel/preset-env` seems to work for me. https://babeljs.io/docs/en/babel-preset-env#modules

```
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "modules": "cjs",
      }
    ],
  ]
}
```

**Comment by Trias** [comment]
Based on @staeke 's [comment](https://github.com/webpack/webpack/issues/3957#issuecomment-470224450) i've created a webpack plugin which will add debug names to generated files (with @staeke permission). 

See https://github.com/Trias/debug-names-webpack-plugin

Happy debugging!
